src/main/java/com/exasol/adapter/document/OnNoMoreElementsIterator.java

```diff
@@ -0,0 +1,56 @@
+package com.exasol.adapter.document;
+
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+
+/**
+ * This {@link Iterator} invokes a method when {@link #hasNext()} returned false the first time or when {@link #next()}
+ * was called but no more elements are available.
+ * <p>
+ * You can use this iterator for example to close a resource.
+ * </p>
+ * 
+ * @param <T> type of the iterator
+ */
+public class OnNoMoreElementsIterator<T> implements Iterator<T> {
```

Comment:
I would suggest to rename this class. But I could not find better name yet.

Suggestions: `FinalizingIterator`, `CallbackAtEndIterator`, `CallbackIterator`